// Copyright Â© 2017-2021 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

syntax = "proto3";

package TW.Avalanche.Proto;
option java_package = "wallet.core.jni.proto";

message SECP256K1TransferOutput {
    uint32 type_id = 1;
    uint64 amount = 2;
    uint64 locktime = 3;
    uint32 threshold = 4;

    // 4 bytes + len(addresses) * 20 bytes
    repeated bytes addresses = 5;
}

message TransactionOutput {
    oneof output {
        SECP256K1TransferOutput secp_transfer_output = 1;
    }
}

message SECP256K1TransferInput {
    uint32 type_id = 1;
    uint64 amount = 2;

    // 4 bytes + len(address_indices) * 4 bytes
    repeated uint32 address_indices = 3;
}

message TransactionInput {
    oneof input {
        SECP256K1TransferInput secp_transfer_input = 1;
    }
}

message TransferableOutput {
    bytes asset_id = 1;  // 32 bytes
    TransactionOutput output = 2;
}

message TransferableInput {
    bytes tx_id = 1;  // 32 bytes
    uint32 utxo_index = 2;
    bytes asset_id = 3;  // 32 bytes
    TransactionInput input = 4;

    // not encoded. represented as pubkeys.
    repeated bytes spendable_addresses = 5;
}

message BaseTx {
    uint32 type_id = 1;
    uint32 network_id = 2;
    bytes blockchain_id = 3;  // 32 bytes

    // 4 bytes + size(outs)
    repeated TransferableOutput outputs = 4;

    // 4 bytes + size(ins)
    repeated TransferableInput inputs = 5;

    // 4 bytes + size(memo)
    bytes memo = 6;
}

message SigningInput {
    repeated bytes private_keys = 1;
    oneof tx_oneof {
        BaseTx base_tx = 2;
    }
}

// Transaction signing output.
message SigningOutput {
    // Signed and encoded transaction bytes.
    bytes encoded = 1;
}
